File: backend/src/main/java/com/tailorstudio/app/service/MePayloadService.java
package com.tailorstudio.app.service;

import com.tailorstudio.app.domain.UserRole;
import com.tailorstudio.app.repo.BusinessRepository;
import com.tailorstudio.app.security.StudioUserDetails;
import org.springframework.stereotype.Service;

import java.util.LinkedHashMap;
import java.util.Map;

/** Shared shape for {@code GET /api/me} and post-login OTP verify (saves one client round trip). */
@Service
public class MePayloadService {

    private final BusinessRepository businessRepository;

    public MePayloadService(BusinessRepository businessRepository) {
        this.businessRepository = businessRepository;
    }

    public Map<String, Object> buildFor(StudioUserDetails u) {
        var business = businessRepository.findById(u.getBusinessId()).orElseThrow();
        Map<String, Object> m = new LinkedHashMap<>();
        m.put("email", u.getUsername());
        m.put("fullName", u.getFullName());
        m.put("role", u.getRole().name());
        m.put("businessId", business.getId());
        m.put("businessName", business.getName());
        m.put("tagline", business.getTagline());
        m.put("address", business.getAddress());
        m.put("phone", business.getPhone());
        m.put("secondaryPhone", business.getSecondaryPhone());
        if (u.getRole() == UserRole.OWNER) {
            m.put("joinCode", business.getJoinCode());
        }
        return m;
    }
}

File: server/src/services/mePayloadService.ts
import { Business } from '../models/Business.js'
import type { UserDoc } from '../models/User.js'
import { parseObjectId } from '../utils/objectId.js'
import { leanOne } from '../utils/lean.js'

export async function buildMe(user: UserDoc) {
  const b = await Business.findById(parseObjectId(user.businessId))
    .select('name tagline address phone secondaryPhone joinCode')
    .lean()
  if (!b) throw new Error('Business not found')
  const m: Record<string, unknown> = {
    email: user.email,
    fullName: user.fullName,
    role: user.role,
    businessId: String(b._id),
    businessName: leanOne<{ name: string }>(b).name,
    tagline: leanOne<{ tagline?: string }>(b).tagline ?? null,
    address: leanOne<{ address?: string }>(b).address ?? null,
    phone: leanOne<{ phone?: string }>(b).phone ?? null,
    secondaryPhone: leanOne<{ secondaryPhone?: string }>(b).secondaryPhone ?? null,
  }
  if (user.role === 'OWNER') m.joinCode = leanOne<{ joinCode: string }>(b).joinCode
  return m
}
